boa/src/builtins/date/mod.rs

```diff
@@ -507,29 +472,79 @@ impl Date {
-        let hour = hour as u32;
-        let min = min as u32;
-        let sec = sec as u32;
-        let milli = milli as u32;
+        if (0.0..=99.0).contains(&year) {
+            year += 1900.0
+        }
 
-        let year = if (0..=99).contains(&year) {
-            1900 + year
+        let mut date = Self(
+            NaiveDateTime::from_timestamp_opt(0, 0)
+                .and_then(|local| ignore_ambiguity(Local.from_local_datetime(&local)))
+                .map(|local| local.naive_utc())
+                .filter(|time| Self::time_clip(time.timestamp_millis() as f64).is_some()),
+        );
+
+        date.set_components(
+            false,
+            Some(year),
+            Some(month),
+            Some(day),
+            Some(hour),
+            Some(min),
+            Some(sec),
+            Some(milli),
+        );
+
+        this.set_data(ObjectData::Date(date));
+
+        Ok(this.clone())
+    }
+
+    /// `Date.prototype[@@toPrimitive]`
+    ///
+    /// The [@@toPrimitive]() method converts a Date object to a primitive value.
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-date.prototype-@@toprimitive
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/@@toPrimitive
+    #[allow(clippy::wrong_self_convention)]
+    pub(crate) fn to_primitive(
+        this: &JsValue,
+        args: &[JsValue],
+        context: &mut Context,
+    ) -> Result<JsValue> {
+        // 1. Let O be the this value.
+        // 2. If Type(O) is not Object, throw a TypeError exception.
+        let o = if let Some(o) = this.as_object() {
+            o
         } else {
-            year
+            return context.throw_type_error("Date.prototype[@@toPrimitive] called on non object");
         };
 
-        let final_date = NaiveDate::from_ymd_opt(year, month + 1, day)
-            .and_then(|naive_date| naive_date.and_hms_milli_opt(hour, min, sec, milli))
-            .and_then(|local| ignore_ambiguity(Local.from_local_datetime(&local)))
-            .map(|local| local.naive_utc())
-            .filter(|time| Self::time_clip(time.timestamp_millis() as f64).is_some());
+        let hint = args
+            .get(0)
+            .cloned()
+            .unwrap_or_default()
+            .to_string(context)?;
+
+        let try_first = match hint.as_str() {
+            // 3. If hint is "string" or "default", then
+            // a. Let tryFirst be string.
+            "string" | "default" => PreferredType::String,
+            // 4. Else if hint is "number", then
+            // a. Let tryFirst be number.
+            "number" => PreferredType::Number,
+            // 5. Else, throw a TypeError exception.
+            _ => {
+                return context
+                    .throw_type_error("Date.prototype[@@toPrimitive] called with invalid hint")
+            }
+        };
```

Comment:
```suggestion
        let hint = args
            .get(0)
            .cloned()
            .unwrap_or_default();

        let try_first = match hint.as_string().map(|x| x.as_str()) {
            // 3. If hint is "string" or "default", then
            // a. Let tryFirst be string.
            Some("string") | Some("default") => PreferredType::String,
            // 4. Else if hint is "number", then
            // a. Let tryFirst be number.
            Some("number") => PreferredType::Number,
            // 5. Else, throw a TypeError exception.
            _ => {
                return context
                    .throw_type_error("Date.prototype[@@toPrimitive] called with invalid hint")
            }
        };
```
I dont think we should call `to_string` since it can have side effects and is not spec complaint.